Add documentation for preparing database for Airflow (#10413)
mistercrunch/Airflow: docs/howto/initialize-database.rst

   want to set a default schema for your role with a
   command similar to ``ALTER ROLE username SET search_path = airflow, foobar;``

Setup your database to host Airflow
-----------------------------------

Create a database called ``airflow`` and a database user that Airflow
will use to access this database.

Example, for **MySQL**:

> .. code-block:: sql

Reviewer: ```suggestion
.. code-block:: sql

```

   CREATE DATABASE airflow CHARACTER SET utf8_mb3 COLLATE utf8_unicode_ci;
   CREATE USER 'airflow' IDENTIFIED BY 'airflow';
   GRANT ALL PRIVILEGES ON airflow.* TO 'airflow';

> 

Reviewer 2: ```suggestion

If you would like to use `utf8mb4` encoding, you need to add an additional configuration option `sql_engine_collation_for_ids` set to `utf8mb3_general_ci` - otherwise index sizes of some Airflow tables will exceed maximum index size allowed in MySQL.

```

I still we need to add this :).

Reviewer 2: My bad with single backticks :(

Example, for **Postgres**:

> .. code-block:: sql

Reviewer: ```suggestion
.. code-block:: sql

```

   CREATE DATABASE airflow;
   CREATE USER airflow WITH PASSWORD 'airflow';
   GRANT ALL PRIVILEGES ON DATABASE airflow TO airflow;

> You may need to update your Postgres `pg_hba.conf` to add the
> `airflow` user to the database access control list; and to reload

Reviewer: ```suggestion
You may need to update your Postgres ``pg_hba.conf`` to add the
``airflow`` user to the database access control list; and to reload
```
rst required double `.

the database configuration to load your change. See

> [The pg_hba.conf File](https://www.postgresql.org/docs/current/auth-pg-hba-conf.html)

Reviewer: ```suggestion
`The pg_hba.conf File <https://www.postgresql.org/docs/current/auth-pg-hba-conf.html>`__
```

in the Postgres documentation to learn more.

Configure Airflow's database connection string
----------------------------------------------

Once you've setup your database to host Airflow, you'll need to alter the
SqlAlchemy connection string located in ``sql_alchemy_conn`` option in ``[core]`` section in your configuration file

> ``$AIRFLOW_HOME/airflow.cfg``.

Reviewer: It would be ideal to mention the configuration about environment variables here, because this option is configured this way, but I can do it in a separate PR.

Author: Okay, thank you very much!  :)

Configure a worker that supports parallelism
--------------------------------------------

> You should then also change the "executor" option in the ``[core]`` option to use "LocalExecutor", an executor that can parallelize task instances locally.

Reviewer: ```suggestion
You should then also change the ``executor`` option in the ``[core]`` section to  ``LocalExecutor``. an executor that can parallelize task instances locally.
```

> You should then also change the "executor" option in the ``[core]`` option to use "LocalExecutor", an executor that can parallelize task instances locally.

Reviewer: ```suggestion
You should then also change the ``executor`` option in the ``[core]`` option to use ``LocalExecutor``, an executor that can parallelize task instances locally.
```

   want to set a default schema for your role with a
   command similar to ``ALTER ROLE username SET search_path = airflow, foobar;``

Setup your database to host Airflow
-----------------------------------

Create a database called ``airflow`` and a database user that Airflow
will use to access this database.

Example, for **MySQL**:

.. code-block:: sql

>    CREATE DATABASE airflow CHARACTER SET utf8 COLLATE utf8_unicode_ci;

Reviewer 2: ```suggestion
   CREATE DATABASE airflow CHARACTER SET utf8_mb3 COLLATE utf8_unicode_ci;
```
Additionally  we must add section that describes how to configure Airflow with mysql `utf8_mb4`. 

Airflow will not work without additional configuration with `utf8_mb4` character set. The index size in few tables is too large for it (maximum index size in 5.7 and 8.0 is 3072 characters and we are exceeding the limit if we use utf8_mb4 - each character in `utf8mb4` is 4 bytes rather than 3 bytes in `utf8_mb3`): https://dev.mysql.com/doc/refman/8.0/en/innodb-limits.html.

This is true for MySQL 5.7 but also remains a problem in MySQL 8, even if we do not officially support it (yet). Additionally: in MySQL 8 `utf8_mb3` is  deprecated and `utf8_mb4` is default when `utf8` is used: https://mysqlserverteam.com/mysql-8-0-when-to-use-utf8mb3-over-utf8mb4/

There is a change I implemented that allows using `utf8mb4` character set but it requires special configuration of Airflow: https://github.com/apache/airflow/pull/7570/files

How it works - you need to set `sql_engine_collation_for_ids` parameter in the configuration of Airflow to `utf8_mb3` if the whole database is `utf8_mb4`. This will switch all the long columns that are used in indexes to use `utf8_mb3`, so the index for those columns does not exceed the maximum index size.

Hopefully, in MySQL 9 (or 10?) they will further increase the index size ¯\_(ツ)_/¯

   CREATE USER 'airflow' IDENTIFIED BY 'airflow';
   GRANT ALL PRIVILEGES ON airflow.* TO 'airflow';

Example, for **Postgres**:

.. code-block:: sql

   CREATE DATABASE airflow;
   CREATE USER airflow WITH PASSWORD 'airflow';
   GRANT ALL PRIVILEGES ON DATABASE airflow TO airflow;

You may need to update your Postgres ``pg_hba.conf`` to add the
``airflow`` user to the database access control list; and to reload
the database configuration to load your change. See
`The pg_hba.conf File <https://www.postgresql.org/docs/current/auth-pg-hba-conf.html>`__
in the Postgres documentation to learn more.

Configure Airflow's database connection string
----------------------------------------------

Once you've setup your database to host Airflow, you'll need to alter the
SqlAlchemy connection string located in ``sql_alchemy_conn`` option in ``[core]`` section in your configuration file
``$AIRFLOW_HOME/airflow.cfg``. 

Configure a worker that supports parallelism
--------------------------------------------

> You should then also change the ``executor`` option in the ``[core]`` option to use ``LocalExecutor``, an executor that can parallelize task instances locally.

Reviewer 2: Maybe 
```suggestion
You should then also change the ``executor`` option in the ``[core]`` option to not use the ``SequentiaExecutor``. Teh ``SequentialExecutor`` cannot parallelize task instance execution. You need to use the `LocalExecutor` if you want to run tasks on single machine or ``CeleryExecutor``, ``KubernetesExecutor`` or ``DaskExecutor`` (those are example ones) if you distribute your tasks between multiple machines.
```

Author: I've taken the suggested edit from @mik-laj , thanks, @potiuk !

   want to set a default schema for your role with a
   command similar to ``ALTER ROLE username SET search_path = airflow, foobar;``

Setup your database to host Airflow
-----------------------------------

Create a database called ``airflow`` and a database user that Airflow
will use to access this database.

Example, for **MySQL**:

.. code-block:: sql
   CREATE DATABASE airflow CHARACTER SET utf8_mb3 COLLATE utf8_unicode_ci;
   CREATE USER 'airflow' IDENTIFIED BY 'airflow';
   GRANT ALL PRIVILEGES ON airflow.* TO 'airflow';

> If you would like to use `utf8mb4` encoding, you need to add an additional configuration option `sql_engine_collation_for_ids` set to `utf8mb3_general_ci` - otherwise index sizes of some Airflow tables will exceed maximum index size allowed in MySQL.

Reviewer 2: ```suggestion
If you would like to use ``utf8mb4`` encoding, you need to add an additional configuration option ``sql_engine_collation_for_ids`` set to ``utf8mb3_general_ci`` - otherwise index sizes of some Airflow tables will exceed maximum index size allowed in MySQL.
```

   want to set a default schema for your role with a
   command similar to ``ALTER ROLE username SET search_path = airflow, foobar;``

Setup your database to host Airflow
-----------------------------------

Create a database called ``airflow`` and a database user that Airflow
will use to access this database.

> Example, for **MySQL**:

Reviewer: Can you also add example for PostgresSQL?

.. code-block:: sql
   CREATE DATABASE airflow CHARACTER SET utf8 COLLATE utf8_unicode_ci;
   CREATE USER 'airflow' IDENTIFIED BY 'airflow';
   GRANT ALL PRIVILEGES ON airflow.* TO 'airflow';

> Configure Airflow's database connection string
> ----------------------------------------------
> 
> Once you've setup your database to host Airflow, you'll need to alter the
> SqlAlchemy connection string located in your configuration file
> ``$AIRFLOW_HOME/airflow.cfg``. You should then also change the "executor"
> setting to use "LocalExecutor", an executor that can parallelize task
> instances locally.

Reviewer: ```suggestion
Configure Airflow's database connection string
----------------------------------------------

Once you've setup your database to host Airflow, you'll need to alter the
SqlAlchemy connection string located in ``sql_alchemy_conn`` option in ``[core]`` section in your configuration file
``$AIRFLOW_HOME/airflow.cfg``. 

Configure a worker that supports parallelism
--------------------------------------------

You should then also change the "executor" option in the ``[core]`` option to use "LocalExecutor", an executor that can parallelize task instances locally.
```
